### PR TITLE
Media: add media section

### PR DESCRIPTION
**Pull request by retrofox**
This PR makes a number of improvements to the media section and updates the media modal to include an action bar at the top. The media section is controlled by the `manage/media` feature flag. This is currently enabled in dev, wpcalypso, horizon and test. The media section can be shown by clicking on the `Media` option into the sidebar.

Related Design Discussion: https://github.com/Automattic/wp-calypso/issues/10805

### The whole picture

image

### Primary Changes
Beyond add the media library this PR implements many changes. Following, the most important ones:

#### Media -> Add item

It allows adding a new item quickly. 

#### Media -> Plan storage nudge

A plan storage nudge is shown in the filter bar is the site doesn't have a business plan.

#### Media -> Item detail dialog

* Added `DONE` and `DELETE` buttons

image

#### Post editor -> add media -> Item detail dialog

* Actions buttons were removed from the bottom
* Added the actions bar above the media gallery 

-----------------------
#### Media Modal
- Navigate to http://calypso.localhost:3000/post/
- Select a site when prompted
- Click on insert content: 
screen shot 2017-03-09 at 12 18 11 pm

- The new action bar appears at the top. 
- The plan storage component in the top right corner can be seen for folks with a wpcom free/personal/premium plan.
screen shot 2017-03-09 at 12 18 27 pm

- Clicking "Add new" opens the file picker
screen shot 2017-03-09 at 12 21 10 pm

- Click on the chevron next to "add new" gives you this option:
screen shot 2017-03-09 at 12 21 43 pm

- Clicking on Add Via Url updates the interface:
screen shot 2017-03-09 at 12 21 43 pm

- Selecting items adds an edit and delete button to the action bar
screen shot 2017-03-09 at 12 22 33 pm

- Search works as expected
- Other media modal flows work as expected

#### Media Section
- Click on the add button of the media section
screen shot 2017-03-09 at 12 37 08 pm

- The media section should open with the file picker ready
- Clicking on the section link outside of "add" should just open the media section
- The media section visually looks ok
screen shot 2017-03-09 at 12 38 15 pm

- We can edit media like we can do in the media modal
- scrolling down, stickies the action bar:
screen shot 2017-03-09 at 12 39 46 pm

#### Site Icon
- We can update a public site icon correctly
- Navigate to http://calypso.localhost:3000/settings/general/
- Select a public site
- Click on "Change" underneath the site icon
- Follow flow, site icon should update

### Known Visual Issues

There are a few known visual issues in the media section that should be addressed after this feature branch lands. It should be safe to land with these quirks since the media section is not enabled in prod yet. We're hoping to land the feature branch quickly to avoid drift and a larger diff.

- Image Editor in Media Section shows too many buttons:
pasted_image_at_2017_03_09_05_46_pm_360
- Visual quirks at smaller sizes with the sticky bar
screen shot 2017-03-09 at 11 42 10 am

### Pull Requests

This process was done working in different patches:

* Media Section: Move Action buttons to top, multiple selection #11589
* Media Section: Add Link #11672
* Media Section: Action bar buttons + card #11560
* Media Section: Move Action buttons to top, multiple selection #11589
* Media Section: tidying secondary actions #11770
* Component: reimplement plan-storage structure using flexbox  #11769
* Media: add plan-storage component #11749
* Plan Storage: Minor style adjustments #11858
* Media Library: fix elevation visual bug in search icon #11836
* Media: fix-plan storage for business plan #11870
* Media modal: do not apply custom styles to uploading-by-URL box #11932
* Media: add `Done` and `Delete` buttons https://github.com/Automattic/wp-calypso/pull/11910
* Media Library: upload-url janitorial #11828 
* Media Library: improve uploading by URL section #11830 
* Media Section: Style cleanup for action bar #11873 

h/t to @iamtakashi @artpi @retrofox @gwwar

### CODE REVIEW

**Comment by gwwar** [comment]
cc @alisterscott @hoverduck would it be possible to run e2e's against this branch? Let us know if we need to update any test suites.

**Comment by hoverduck** [comment]
@gwwar The e2e tests look good except (unsurprisingly) the [media upload test](https://github.com/Automattic/wp-e2e-tests/blob/master/specs/wp-media-upload-spec.js) needs some updates.  Since you're the original author can you take a look?  If not I can get to it tomorrow

File: client/my-sites/media-library/filter-bar.jsx [hunk 1]
```diff
@@ -97,23 +99,44 @@ export class MediaLibraryFilterBar extends Component {
...
+				delaySearch={ true } />
+		);
+	}
+
+	renderPlanStorage() {
+		const eventName = 'calypso_upgrade_nudge_impression';
+		const eventProperties = { cta_name: 'plan-media-storage' };
+		return (
+			<PlanStorage siteId={ this.props.site.ID }>
+				<TrackComponentView eventName={ eventName } eventProperties={ eventProperties } />
```
Comment (dmsnell) [inline client/my-sites/media-library/filter-bar.jsx]: fun fact! when I'm just passing named properties like this I find the spread syntax helpful…

```js
<TrackComponentView { ...{
	eventName,
	eventProperties
} } />
```

**Comment by gwwar** [comment]
Thanks @hoverduck I'll try to take a look at that today

File: client/my-sites/media-library/upload-url.jsx [hunk 2]
```diff
@@ -13,71 +14,70 @@ import analytics from 'lib/analytics';
...
+		isError: false,
+	};
+
+	upload = event => {
+		const isError = ! event.target.checkValidity();
+
+		this.setState( { isError } );
 
 		if ( isError || ! this.props.site ) {
+			event.preventDefault();
```
Comment (klimeryk) [inline client/my-sites/media-library/upload-url.jsx]: Just a quick drive-by comment - it seems we need to call `event.preventDefault()` in both cases, so it should probably be called once (before this if branch) - otherwise someone might think that one of these calls should not be there :)

File: client/my-sites/media/main.jsx [hunk 3]
```diff
@@ -97,54 +115,141 @@ export default React.createClass( {
...
+		const selectedCount = selected.length;
+		const confirmMessage = this.translate(
+			'Are you sure you want to permanently delete this item?',
+			'Are you sure you want to permanently delete these items?',
+			{ count: selectedCount }
+		);
+
+		accept( confirmMessage, accepted => {
+			if ( ! accepted ) {
+				return null;
```
Comment (klimeryk) [inline client/my-sites/media/main.jsx]: Seems like just `return;` should suffice :)

File: client/my-sites/media-library/upload-url.jsx (hunk 2 above)
Comment (retrofox) [inline client/my-sites/media-library/upload-url.jsx]: I'm sorry Igor, I don't understand completely what do you mean. Could you point straight in the file of this PR?

File: client/my-sites/media/main.jsx (hunk 3 above)
Comment (retrofox) [inline client/my-sites/media/main.jsx]: Yes, agree. But I've gotten feedback from other PR that we should be explicit in this sense. Although `return` returns a `null` if nothing is defined, it's more understandable when the `null` value is defined explicitly.

File: client/my-sites/media-library/upload-url.jsx (hunk 2 above)
Comment (retrofox) [inline client/my-sites/media-library/upload-url.jsx]: Now I got it. You're right. Thanks.

Comment (klimeryk) [inline client/my-sites/media-library/upload-url.jsx]: Ah, sorry. So, no matter which flow we take in this function we end up calling `event.preventDefault`. Either here: https://github.com/Automattic/wp-calypso/pull/11822/files#diff-246b58e6f07d7dce525ca439d82eccf9R41 or here: https://github.com/Automattic/wp-calypso/pull/11822/files#diff-246b58e6f07d7dce525ca439d82eccf9R45.
So I thought that this call should be "deduplicated" and moved to a single call, somewhere above https://github.com/Automattic/wp-calypso/pull/11822/files#diff-246b58e6f07d7dce525ca439d82eccf9R40. Less chances of one of these calls getting deleted and it shows the intent (always prevent default behaviour of the event) better.

**Comment by alisterscott** [comment]
> Thanks @hoverduck I'll try to take a look at that today

@gwwar: if you don't find time please LMK and I can do this for you

File: client/my-sites/media/main.jsx (hunk 3 above)
Comment (klimeryk) [inline client/my-sites/media/main.jsx]: Ah, got it - I was thrown off by no explicit return in the success scenario, but if that's intended then cool 👍

Comment (retrofox) [inline client/my-sites/media/main.jsx]: yes ... we aren't being consistents with this. :-|

File: client/my-sites/media-library/upload-url.jsx (hunk 2 above)
Comment (retrofox) [inline client/my-sites/media-library/upload-url.jsx]: Actually, this line seems to be unneeded. Good one 👍

File: client/my-sites/media/main.jsx (hunk 3 above)
Comment (dmsnell) [inline client/my-sites/media/main.jsx]: > Although return returns a null if nothing is defined, it's more understandable when the null value is defined explicitly.

interesting…what reason was given to be explicit in the return value there?

also this is wrong. `return` returns `undefined` if no value is given.

Comment (retrofox) [inline client/my-sites/media/main.jsx]: > also this is wrong. return returns undefined if no value is given.

You're right. Another reason to return `null` ;-)

Comment (dmsnell) [inline client/my-sites/media/main.jsx]: > You're right. Another reason to return null ;-)

the way I typically do it is that if my only reason to use `return` is to early-abort from a function then I'll just `return` because that's how it would have happened if the function had returned normally given no explicit value.

if, on the other hand, the function is designed to return a value and something else is counting on it I'll want to explicitly return something. bonus points if it's the same data type as would otherwise return.

for example, an empty string could be more appropriate for a string function than `null`, a `0` might be more appropriate for a number function, an empty list for a list function, etc…

but if I'm calling a function for its side-effects only and don't expect a value, I think getting _something_ back could also be confusing

**Comment by gwwar** [comment]
Thanks for the offer @alisterscott ❤️  Tests should be fixed in https://github.com/Automattic/wp-e2e-tests/pull/412

File: client/my-sites/media/main.jsx (hunk 3 above)
Comment (gwwar) [inline client/my-sites/media/main.jsx]: > But I've gotten feedback from other PR that we should be explicit in this sense. Although return returns a null if nothing is defined, it's more understandable when the null value is defined explicitly.

Do you have the link @retrofox? We might have gotten confused that this was a render helper. A `return null` for rending is an explicit indication for React that we want to render nothing.

I missed this on the original PR, but it doesn't look like anything uses this return value, in which case we should just `return;` I agree with @dmsnell on consistency. A function should either always return some value if we're using it, or nothing if we don't.

Comment (retrofox) [inline client/my-sites/media/main.jsx]: Yes, sorry. Last night I forgot to leave my answer here. 
I was confusing with the render method of React component. Let me update the code. I'm sorry again and thanks!

Comment (retrofox) [inline client/my-sites/media/main.jsx]: rebased and updated

**Comment by folletto** [comment]
Ok I gave a round of review. Really solid work, this is now feeling very nice :)

Some feedback:

#### (1) Media -> Actions bar

Would be ideal if the popup text wasn't small and uppercase, but normal and with a standard font (`font-size: 14px`, not uppercase). Also should be a bit taller (`26px` → `33px`). See the ••• menus in Pages and Themes for reference.

screen shot 2017-03-10 at 14 47 50


#### (2) Media -> Plan storage nudge 

I'd leave the final decision to @iamtakashi, but I think the whole area should be clickable, not just "Upgrade", and it should also have a hover effect to signify it's all clickable (i.e. the whole block gets a slight gray background and the text gets colored).

screen shot 2017-03-10 at 14 50 58


#### (3) Media -> Item detail dialog

I'd also add a "Cancel" / "Close" button. I was on the edge when we discussed  this with @iamtakashi, but now seeing it live, the top arrow to go back doesn't seem very intuitive, I'd even remove it leaving just the buttons. This for two reasons:

1. The primary button, and the attention, is down there already.
2. (minor) Follows our dialog patterns.

This is a minor detail, so skip it freely if you feel doesn't belong to this PR.

**Comment by gwwar** [comment]
Thanks @folletto for testing!

> (1) Media -> Actions bar Would be ideal if the popup text wasn't small and uppercase, but normal and with a standard font 
@retrofox let's update styling for this

> (2) Media -> Plan storage nudge I'd leave the final decision to @iamtakashi, but I think the whole area should be clickable, not just "Upgrade" (3) Media -> Item detail dialog I'd also add a "Cancel" / "Close" button. 

I think we can follow up pretty quickly and discuss points 2 and 3 after the feature branch lands.

**Comment by retrofox** [comment]
> > (1) Media -> Actions bar Would be ideal if the popup text wasn't small and uppercase, but normal and with a standard font

> @retrofox let's update styling for this

sure.

Thanks @folletto for the feedback. Let's create and add  issues for (2) and (3) in our backlog as well.

**Comment by retrofox** [comment]
f5061c2

**Comment by folletto** [comment]
It's a nitpick, but i'd make the clickable item `33px` tall too. :)

**Comment by retrofox** [comment]
> It's a nitpick

nop, it isn't.

> but i'd make the clickable item 33px tall too. :)

Sorry, do you mean increase from `26px` to `33px` there?

**Comment by folletto** [comment]
> Sorry, do you mean increase from 26px to 33px there?

Yep. The blue area basically.

screen shot 2017-03-10 at 17 10 02

**Comment by retrofox** [comment]
> > Sorry, do you mean increase from 26px to 33px there?

> Yep. The blue area basically.

yes, already done. I've removed the `is-compact` className from the `MenuItem` component.
